test/github-issues/11421/issue-11421.ts

```diff
@@ -0,0 +1,65 @@
+import "reflect-metadata"
+import {
+    createTestingConnections,
+    closeTestingConnections,
+} from "../../utils/test-utils"
+import { DataSource } from "../../../src/index.js"
+import { expect } from "chai"
+import { Book } from "./entity/Book"
+
+describe("github issues > #11421 findOne returns null if id is not selected", () => {
+    let dataSources: DataSource[]
+
+    before(
+        async () =>
+            (dataSources = await createTestingConnections({
+                entities: [__dirname + "/entity/*{.js,.ts}"],
+                schemaCreate: true,
+                dropSchema: true,
+                enabledDrivers: ["postgres"],
+                logging: true,
+            })),
+    )
+
+    after(() => closeTestingConnections(dataSources))
+
+    it("returns an entity when selecting only a column with a value", () =>
+        Promise.all(
+            dataSources.map(async (dataSource) => {
+                const repository = dataSource.getRepository(Book)
+
+                let book = repository.create({ name: "Lord of the Rings" })
+                book = await repository.save(book, { reload: true })
+
+                expect(book.id).to.be.a.string
+
+                const retrieved = await repository.findOne({
+                    where: { id: book.id },
+                    select: { name: true },
+                })
+
+                expect(retrieved).to.be.instanceOf(Book)
+                expect(retrieved!.name).to.eq("Lord of the Rings")
+            }),
+        ))
+
+    it("returns an entity when selecting only a column with a null", () =>
+        Promise.all(
+            dataSources.map(async (dataSource) => {
+                const repository = dataSource.getRepository(Book)
+
+                let book = repository.create({ name: "Lord of the Rings" })
+                book = await repository.save(book, { reload: true })
+
+                expect(book.id).to.be.a.string
+
+                const retrieved = await repository.findOne({
+                    where: { id: book.id },
+                    select: { author: true },
+                })
+
+                expect(retrieved).to.be.instanceOf(Book)
```

Comment:
I debugged the code and here is the piece of code that creates the [issue](https://github.com/typeorm/typeorm/blob/24a03696731ab212903f2173eacf586413da5d17/src/query-builder/transformer/RawSqlResultsToEntityTransformer.ts#L208)

In this case we are selecting a column with null value and `hasColumns` is false. The result of this function is undefined.
@alumni @naorpeled What do you think about it?
Should we add a fallback?

Reply:
I remember that part from before the the performance improvement PR - I believe it was added for relations, i.e. when the related entity is selected, it will be left joined; if there are no results, we'd get null for all the columns of the related entity.